File: scripts/python/src/tools/utils/sysinfo/hardware.py
import re

from tools.utils.sysinfo.branding import resolve_brand, strip_brand
from tools.utils.sysinfo.devices import cache_size, gpu_memory, named_gpu, nvidia_for, swap_totals
from tools.utils.sysinfo.facts import fact, facts
from tools.utils.sysinfo.formatting import (
    as_dict,
    as_list,
    average,
    capacity,
    format_bytes,
    format_frequency,
    format_temperature,
    join_parts,
    memory_capacity,
    memory_summary,
    percentage,
)
from tools.utils.sysinfo.models import Component
from tools.utils.sysinfo.normalization import (
    is_macos,
    is_virtual_disk,
    text_value,
    useful_device_name,
)
from tools.utils.sysinfo.profiles import (
    CPU_PROFILES,
    DISK_PROFILES,
    GPU_PROFILES,
    profile_for,
)


def cpu_component(snapshot):
    cpu = as_dict(snapshot.result("CPU", {}))
    name = cpu.get("cpu") or "Unknown CPU"
    vendor = cpu.get("vendor") or ""
    brand = resolve_brand("cpu", vendor, name)
    profile = profile_for(name, CPU_PROFILES)
    cores = as_dict(cpu.get("cores"))
    frequency = as_dict(cpu.get("frequency"))
    cache = as_dict(snapshot.result("CPUCache", {}))
    usage = average(as_list(snapshot.result("CPUUsage", [])))
    threads = ""
    if cores.get("physical") and cores.get("logical"):
        threads = f"{cores['physical']} cores / {cores['logical']} threads"
    l3 = cache_size(cache.get("l3"))
    family = profile.family or cpu.get("codeName") or ""
    return Component(
        kind="cpu",
        label="CPU",
        vendor=vendor or brand.name,
        model=strip_brand(name, brand),
        identifiers=(name, family),
        facts=facts(
            fact("Cores", threads),
            fact("Cache", f"{l3} L3" if l3 else ""),
            fact("Family", family),
            fact("Clock", format_frequency(frequency.get("max"))),
            fact("Power", f"{profile.tdp} W TDP" if profile.tdp else ""),
            fact("Temperature", format_temperature(cpu.get("temperature"))),
            fact("Load", f"{usage:.0f}%" if usage is not None else ""),
            fact("Process ISA", cpu.get("march") or ""),
            fact("Process", cpu.get("technology") or ""),
        ),
    )


def gpu_components(snapshot):
    components = []
    gpus = as_list(snapshot.result("GPU", []))
    discrete_present = any(gpu.get("type") == "Discrete" for gpu in gpus)
    for gpu in gpus:
        name = named_gpu(gpu)
        vendor = gpu.get("vendor") or ""
        brand = resolve_brand("gpu", vendor, name)
        profile = profile_for(name, GPU_PROFILES)
        nvidia = nvidia_for(snapshot, gpu)
        used, total = gpu_memory(gpu, nvidia)
        temperature = nvidia.get("temperature")
        if temperature is None:
            temperature = gpu.get("temperature")
        utilization = nvidia.get("utilization")
        if utilization is None:
            utilization = gpu.get("coreUsage")
        clock = nvidia.get("clock_mhz")
        if clock is None:
            clock = gpu.get("frequency")
        power_draw = nvidia.get("power_draw")
        power_limit = nvidia.get("power_limit")
        power = ""
        if power_draw is not None:
            power = f"{power_draw:.0f} W"
            if power_limit:
                power += f" / {power_limit:.0f} W"
        pcie = as_dict(gpu.get("pcieSpeed"))
        pcie_max = as_dict(pcie.get("max"))
        pcie_value = ""
        if pcie_max.get("gen") and pcie_max.get("lanes"):
            pcie_value = f"PCIe {pcie_max['gen']}.0 ×{pcie_max['lanes']}"
        driver = nvidia.get("driver") or gpu.get("driver") or ""
        integrated = gpu.get("type") == "Integrated"
        memory = profile.memory or (f"{format_bytes(total)} VRAM" if total else "")
        components.append(
            Component(
                kind="gpu",
                label="INTEGRATED GPU" if integrated else "GPU",
                vendor=vendor or brand.name,
                model=strip_brand(name, brand),
                identifiers=(name, gpu.get("driver") or ""),
                facts=facts(
                    fact("Architecture", profile.architecture),
                    fact("Memory", memory),
                    fact("Cores", profile.cores),
                    fact(
                        "VRAM use", f"{format_bytes(used)} / {format_bytes(total)}" if total else ""
                    ),
                    fact("Load", f"{utilization:.0f}%" if utilization is not None else ""),
                    fact("Clock", format_frequency(clock)),
                    fact("Temperature", format_temperature(temperature)),
                    fact("Power", power),
                    fact("Link", pcie_value),
                    fact("Driver", driver),
                ),
                compact=not integrated or not discrete_present,
            )
        )
    return components


def physical_memory_hints(snapshot):
    values = []
    for module in as_list(snapshot.result("PhysicalMemory", [])):
        for key in ("vendor", "manufacturer", "partNumber", "type"):
            if module.get(key):
                values.append(str(module[key]))
    return tuple(values)


def memory_component(snapshot):
    memory = as_dict(snapshot.result("Memory", {}))
    description = snapshot.hardware.get("memory", "")
    detected = memory.get("total") or 0
    hints = physical_memory_hints(snapshot)
    macos = is_macos(snapshot)
    brand = resolve_brand("memory", description, *hints, "Apple" if macos else "")
    kit_match = re.search(r"\b[A-Z0-9]{8,}\b", description)
    used = memory.get("used") or 0
    swap_used, swap_total = swap_totals(snapshot.result("Swap", []))
    module_count = 0 if macos else len(as_list(snapshot.result("PhysicalMemory", [])))
    swap = ""
    if swap_total:
        swap = f"{format_bytes(swap_used)} / {format_bytes(swap_total)}"
    elif not macos:
        swap = "Disabled"
    return Component(
        kind="memory",
        label="MEMORY",
        vendor=brand.name,
        model=memory_summary(description, detected),
        identifiers=(description, *hints),
        facts=facts(
            fact("Kit", kit_match.group(0) if kit_match else ""),
            fact("Modules", str(module_count) if module_count else ""),
            fact("Detected", memory_capacity(detected)),
            fact("Usage", f"{format_bytes(used)} / {format_bytes(detected)}" if detected else ""),
            fact("Load", f"{percentage(used, detected):.0f}%" if detected else ""),
            fact("Swap", swap),
        ),
    )


def motherboard_component(snapshot):
    board = as_dict(snapshot.result("Board", {}))
    configured = snapshot.hardware.get("motherboard", "")
    vendor = board.get("vendor") or ""
    model = configured or board.get("name") or "Unknown motherboard"
    brand = resolve_brand("motherboard", vendor, model)
    return Component(
        kind="motherboard",
        label="MOTHERBOARD",
        vendor=vendor or brand.name,
        model=strip_brand(model, brand),
        identifiers=(vendor, model),
        facts=facts(
            fact("Revision", board.get("version") or ""),
        ),
    )


def disk_components(snapshot):
    components = []
    for disk in as_list(snapshot.result("PhysicalDisk", [])):
        if is_virtual_disk(disk):
            continue
        name = disk.get("name") or "Unknown disk"
        profile = profile_for(name, DISK_PROFILES)
        brand = resolve_brand("storage", name)
        model = strip_brand(profile.family, brand) if profile.family else ""
        model = model or strip_brand(name.removeprefix("ATA "), brand)
        components.append(
            Component(
                kind="storage",
                label="STORAGE",
                vendor=brand.name,
                model=join_parts([model, capacity(disk.get("size")), disk.get("kind")]),
                identifiers=(name, profile.family),
                facts=facts(
                    fact("Device", name.removeprefix("ATA ")),
                    fact("Capacity", capacity(disk.get("size"))),
                    fact("Interface", disk.get("interconnect") or ""),
                    fact("Type", disk.get("kind") or ""),
                    fact("Temperature", format_temperature(disk.get("temperature"))),
                ),
            )
        )
    if components:
        return components
    total = sum(
        as_dict(disk.get("bytes")).get("total") or 0
        for disk in as_list(snapshot.result("Disk", []))
    )
    if not total:
        return []
    return [
        Component(
            kind="storage",
            label="STORAGE",
            vendor="",
            model=capacity(total),
        )
    ]


def configured_components(snapshot):
    values = (
        ("cooling", "CPU COOLING", snapshot.hardware.get("cpu_cooler", "")),
        ("case", "CHASSIS", snapshot.hardware.get("case", "")),
        ("power", "POWER SUPPLY", snapshot.hardware.get("power_supply", "")),
    )
    components = []
    for kind, label, description in values:
        if not description or description == "not set":
            continue
        brand = resolve_brand(kind, description)
        components.append(
            Component(
                kind=kind,
                label=label,
                vendor=brand.name,
                model=strip_brand(description, brand),
                identifiers=(description,),
                compact=False,
            )
        )
    return components


def portable_power_components(snapshot):
    components = []
    macos = is_macos(snapshot)
    for battery in as_list(snapshot.result("Battery", [])):
        raw_name = battery.get("modelName") or battery.get("name")
        name = useful_device_name(raw_name, "Internal battery", ("bq", "smc"))
        vendor = battery.get("manufacturer") or ("Apple" if macos else "")
        components.append(
            Component(
                kind="power",
                label="BATTERY",
                vendor=vendor,
                model=name,
                art_kind="battery",
                identifiers=(vendor, name),
                facts=facts(
                    fact(
                        "Charge",
                        f"{battery.get('capacity'):.0f}%"
                        if isinstance(battery.get("capacity"), (int, float))
                        else "",
                    ),
                    fact("Status", text_value(battery.get("status"))),
                    fact("Temperature", format_temperature(battery.get("temperature"))),
                    fact(
                        "Cycles",
                        str(battery["cycleCount"]) if battery.get("cycleCount") is not None else "",
                    ),
                ),
            )
        )
    for adapter in as_list(snapshot.result("PowerAdapter", [])):
        watts = adapter.get("watts")
        fallback = f"{watts} W" if watts else "Connected"
        name = useful_device_name(adapter.get("modelName") or adapter.get("name"), fallback)
        vendor = adapter.get("manufacturer") or ("Apple" if macos else "")
        output = f"{watts} W" if watts and name != f"{watts} W" else ""
        components.append(
            Component(
                kind="power",
                label="POWER ADAPTER",
                vendor=vendor,
                model=name,
                art_kind="adapter",
                identifiers=(vendor, name),
                facts=facts(
                    fact("Output", output),
                ),
                compact=False,
            )
        )
    return components


def hardware_components(snapshot):
    return (
        cpu_component(snapshot),
        *gpu_components(snapshot),
        memory_component(snapshot),
        motherboard_component(snapshot),
        *disk_components(snapshot),
        *configured_components(snapshot),
        *portable_power_components(snapshot),
    )

File: scripts/rust/crates/sysinfo/src/presentation/hardware.rs
use super::{
    branding::{resolve_brand, strip_brand},
    facts, joined,
};
use crate::formatting::*;
use crate::model::{Component, Snapshot};
use regex::Regex;
use serde_json::Value;
use std::sync::LazyLock;

fn cpu_component(snapshot: &Snapshot) -> Component {
    let cpu = snapshot.result("CPU");
    let name = cpu["cpu"]
        .as_str()
        .filter(|s| !s.is_empty())
        .unwrap_or("Unknown CPU");
    let vendor = string(cpu, "vendor");
    let brand = resolve_brand("cpu", &[vendor, name]);
    let profile = name.to_lowercase().contains("ryzen 7 9800x3d");
    let family = if profile {
        "Granite Ridge X3D"
    } else {
        string(cpu, "codeName")
    };
    let cores = &cpu["cores"];
    let physical = number(&cores["physical"]);
    let logical = number(&cores["logical"]);
    let threads = if physical > 0.0 && logical > 0.0 {
        format!("{physical:.0} cores / {logical:.0} threads")
    } else {
        String::new()
    };
    let cache = list(&snapshot.result("CPUCache")["l3"])
        .iter()
        .map(|entry| {
            number(&entry["size"]) * entry["num"].as_f64().filter(|v| *v != 0.0).unwrap_or(1.0)
        })
        .sum::<f64>();
    let loads = list(snapshot.result("CPUUsage"))
        .iter()
        .filter_map(Value::as_f64)
        .collect::<Vec<_>>();
    let usage = if loads.is_empty() {
        String::new()
    } else {
        format!("{:.0}%", loads.iter().sum::<f64>() / loads.len() as f64)
    };
    Component {
        kind: "cpu".into(),
        label: "CPU".into(),
        vendor: if vendor.is_empty() {
            brand.name.clone()
        } else {
            vendor.into()
        },
        model: strip_brand(name, brand),
        identifiers: vec![name.into(), family.into()],
        compact: true,
        facts: facts([
            ("Cores", threads),
            (
                "Cache",
                if cache > 0.0 {
                    format!("{} L3", format_bytes(cache))
                } else {
                    String::new()
                },
            ),
            ("Family", family.into()),
            ("Clock", format_frequency(number(&cpu["frequency"]["max"]))),
            (
                "Power",
                if profile {
                    "120 W TDP".into()
                } else {
                    String::new()
                },
            ),
            (
                "Temperature",
                format_temperature(cpu["temperature"].as_f64()),
            ),
            ("Load", usage),
            ("Process ISA", text(&cpu["march"])),
            ("Process", text(&cpu["technology"])),
        ]),
        ..Component::default()
    }
}
fn gpu_components(snapshot: &Snapshot) -> Vec<Component> {
    let gpus = list(snapshot.result("GPU"));
    let discrete = gpus.iter().any(|g| g["type"] == "Discrete");
    gpus.iter()
        .map(|gpu| {
            let name = named_gpu(gpu);
            let vendor = string(gpu, "vendor");
            let brand = resolve_brand("gpu", &[vendor, &name]);
            let profile = name.to_lowercase().contains("geforce rtx 5070 ti");
            let nvidia = nvidia_for(snapshot, gpu);
            let (used, total) = gpu_memory(gpu, nvidia);
            let number_from = |native: &str, fallback: &str| {
                nvidia[native].as_f64().or_else(|| gpu[fallback].as_f64())
            };
            let power = nvidia["power_draw"]
                .as_f64()
                .map(
                    |draw| match nvidia["power_limit"].as_f64().filter(|v| *v != 0.0) {
                        Some(limit) => format!("{draw:.0} W / {limit:.0} W"),
                        None => format!("{draw:.0} W"),
                    },
                )
                .unwrap_or_default();
            let pcie = &gpu["pcieSpeed"]["max"];
            let link = if number(&pcie["gen"]) > 0.0 && number(&pcie["lanes"]) > 0.0 {
                format!("PCIe {}.0 ×{}", text(&pcie["gen"]), text(&pcie["lanes"]))
            } else {
                String::new()
            };
            let integrated = gpu["type"] == "Integrated";
            Component {
                kind: "gpu".into(),
                label: if integrated { "INTEGRATED GPU" } else { "GPU" }.into(),
                vendor: if vendor.is_empty() {
                    brand.name.clone()
                } else {
                    vendor.into()
                },
                model: strip_brand(&name, brand),
                identifiers: vec![name, text(&gpu["driver"])],
                compact: !integrated || !discrete,
                facts: facts([
                    (
                        "Architecture",
                        if profile {
                            "Blackwell".into()
                        } else {
                            String::new()
                        },
                    ),
                    (
                        "Memory",
                        if profile {
                            "16 GB GDDR7".into()
                        } else if total > 0.0 {
                            format!("{} VRAM", format_bytes(total))
                        } else {
                            String::new()
                        },
                    ),
                    (
                        "Cores",
                        if profile {
                            "8,960 CUDA cores".into()
                        } else {
                            String::new()
                        },
                    ),
                    (
                        "VRAM use",
                        if total > 0.0 {
                            format!("{} / {}", format_bytes(used), format_bytes(total))
                        } else {
                            String::new()
                        },
                    ),
                    (
                        "Load",
                        number_from("utilization", "coreUsage")
                            .map(|v| format!("{v:.0}%"))
                            .unwrap_or_default(),
                    ),
                    (
                        "Clock",
                        format_frequency(number_from("clock_mhz", "frequency").unwrap_or(0.0)),
                    ),
                    (
                        "Temperature",
                        format_temperature(number_from("temperature", "temperature")),
                    ),
                    ("Power", power),
                    ("Link", link),
                    (
                        "Driver",
                        nvidia["driver"]
                            .as_str()
                            .filter(|v| !v.is_empty())
                            .unwrap_or(string(gpu, "driver"))
                            .into(),
                    ),
                ]),
                ..Component::default()
            }
        })
        .collect()
}
fn memory_component(snapshot: &Snapshot) -> Component {
    static KIT: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\b[A-Z0-9]{8,}\b").unwrap());
    let memory = snapshot.result("Memory");
    let description = snapshot.configured("memory");
    let detected = number(&memory["total"]);
    let used = number(&memory["used"]);
    let modules = list(snapshot.result("PhysicalMemory"));
    let mut hints = vec![description.to_string()];
    for module in modules {
        for key in ["vendor", "manufacturer", "partNumber", "type"] {
            let value = text(&module[key]);
            if !value.is_empty() {
                hints.push(value);
            }
        }
    }
    let mut identities = hints.iter().map(String::as_str).collect::<Vec<_>>();
    if snapshot.is_macos() {
        identities.push("Apple");
    }
    let brand = resolve_brand("memory", &identities);
    let (swap_used, swap_total) =
        list(snapshot.result("Swap"))
            .iter()
            .fold((0.0, 0.0), |(used, total), swap| {
                let bytes = &swap["bytes"];
                (
                    used + number(&bytes["used"]),
                    total + number(&bytes["total"]),
                )
            });
    let swap = if swap_total > 0.0 {
        format!("{} / {}", format_bytes(swap_used), format_bytes(swap_total))
    } else if !snapshot.is_macos() {
        "Disabled".into()
    } else {
        String::new()
    };
    Component {
        kind: "memory".into(),
        label: "MEMORY".into(),
        vendor: brand.name.clone(),
        model: memory_summary(description, detected),
        identifiers: hints,
        compact: true,
        facts: facts([
            (
                "Kit",
                KIT.find(description)
                    .map(|m| m.as_str())
                    .unwrap_or("")
                    .into(),
            ),
            (
                "Modules",
                if !snapshot.is_macos() && !modules.is_empty() {
                    modules.len().to_string()
                } else {
                    String::new()
                },
            ),
            ("Detected", memory_capacity(detected)),
            (
                "Usage",
                if detected > 0.0 {
                    format!("{} / {}", format_bytes(used), format_bytes(detected))
                } else {
                    String::new()
                },
            ),
            (
                "Load",
                if detected > 0.0 {
                    format!("{:.0}%", percentage(used, detected))
                } else {
                    String::new()
                },
            ),
            ("Swap", swap),
        ]),
        ..Component::default()
    }
}
fn motherboard_component(snapshot: &Snapshot) -> Component {
    let board = snapshot.result("Board");
    let configured = snapshot.configured("motherboard");
    let vendor = string(board, "vendor");
    let name = if !configured.is_empty() {
        configured
    } else {
        board["name"]
            .as_str()
            .filter(|v| !v.is_empty())
            .unwrap_or("Unknown motherboard")
    };
    let brand = resolve_brand("motherboard", &[vendor, name]);
    Component {
        kind: "motherboard".into(),
        label: "MOTHERBOARD".into(),
        vendor: if vendor.is_empty() {
            brand.name.clone()
        } else {
            vendor.into()
        },
        model: strip_brand(name, brand),
        identifiers: vec![vendor.into(), name.into()],
        facts: facts([("Revision", text(&board["version"]))]),
        compact: true,
        ..Component::default()
    }
}
fn disk_components(snapshot: &Snapshot) -> Vec<Component> {
    let mut components = Vec::new();
    for disk in list(snapshot.result("PhysicalDisk"))
        .iter()
        .filter(|d| !is_virtual_disk(d))
    {
        let name = disk["name"]
            .as_str()
            .filter(|v| !v.is_empty())
            .unwrap_or("Unknown disk");
        let lower = name.to_lowercase();
        let family = if lower.contains("kingston snvs2000g") {
            "Kingston NV1"
        } else if lower.contains("wdc wd20ezrz") {
            "WD Blue"
        } else {
            ""
        };
        let brand = resolve_brand("storage", &[name]);
        let raw = name.strip_prefix("ATA ").unwrap_or(name);
        let model = strip_brand(if family.is_empty() { raw } else { family }, brand);
        let capacity = capacity(number(&disk["size"]));
        components.push(Component {
            kind: "storage".into(),
            label: "STORAGE".into(),
            vendor: brand.name.clone(),
            model: joined(&[&model, &capacity, string(disk, "kind")]),
            identifiers: vec![name.into(), family.into()],
            compact: true,
            facts: facts([
                ("Device", raw.into()),
                ("Capacity", capacity),
                ("Interface", text(&disk["interconnect"])),
                ("Type", text(&disk["kind"])),
                (
                    "Temperature",
                    format_temperature(disk["temperature"].as_f64()),
                ),
            ]),
            ..Component::default()
        });
    }
    if components.is_empty() {
        let total = list(snapshot.result("Disk"))
            .iter()
            .map(|d| number(&d["bytes"]["total"]))
            .sum::<f64>();
        if total > 0.0 {
            components.push(Component {
                kind: "storage".into(),
                label: "STORAGE".into(),
                model: capacity(total),
                compact: true,
                ..Component::default()
            });
        }
    }
    components
}
fn useful_name(value: &Value, fallback: &str, rejected: &[&str]) -> String {
    let name = text(value).trim().to_owned();
    let lower = name.to_lowercase();
    if name.is_empty()
        || name.chars().all(|c| c.is_ascii_digit())
        || ["unknown", "n/a", "none", "null"].contains(&lower.as_str())
        || rejected.iter().any(|p| lower.starts_with(p))
    {
        fallback.into()
    } else {
        name
    }
}
fn portable_power_components(snapshot: &Snapshot) -> Vec<Component> {
    let mut components = Vec::new();
    for battery in list(snapshot.result("Battery")) {
        let raw = if !text(&battery["modelName"]).is_empty() {
            &battery["modelName"]
        } else {
            &battery["name"]
        };
        let name = useful_name(raw, "Internal battery", &["bq", "smc"]);
        let vendor = battery["manufacturer"]
            .as_str()
            .filter(|v| !v.is_empty())
            .unwrap_or(if snapshot.is_macos() { "Apple" } else { "" });
        components.push(Component {
            kind: "power".into(),
            label: "BATTERY".into(),
            vendor: vendor.into(),
            model: name.clone(),
            art_kind: "battery".into(),
            identifiers: vec![vendor.into(), name],
            compact: true,
            facts: facts([
                (
                    "Charge",
                    battery["capacity"]
                        .as_f64()
                        .map(|v| format!("{v:.0}%"))
                        .unwrap_or_default(),
                ),
                ("Status", text(&battery["status"])),
                (
                    "Temperature",
                    format_temperature(battery["temperature"].as_f64()),
                ),
                ("Cycles", text(&battery["cycleCount"])),
            ]),
        });
    }
    for adapter in list(snapshot.result("PowerAdapter")) {
        let watts = number(&adapter["watts"]);
        let fallback = if watts != 0.0 {
            format!("{watts} W")
        } else {
            "Connected".into()
        };
        let raw = if !text(&adapter["modelName"]).is_empty() {
            &adapter["modelName"]
        } else {
            &adapter["name"]
        };
        let name = useful_name(raw, &fallback, &[]);
        let vendor = adapter["manufacturer"]
            .as_str()
            .filter(|v| !v.is_empty())
            .unwrap_or(if snapshot.is_macos() { "Apple" } else { "" });
        components.push(Component {
            kind: "power".into(),
            label: "POWER ADAPTER".into(),
            vendor: vendor.into(),
            model: name.clone(),
            art_kind: "adapter".into(),
            identifiers: vec![vendor.into(), name.clone()],
            compact: false,
            facts: facts([(
                "Output",
                if watts != 0.0 && name != fallback {
                    fallback
                } else {
                    String::new()
                },
            )]),
        });
    }
    components
}
pub fn hardware_components(snapshot: &Snapshot) -> Vec<Component> {
    let mut components = vec![cpu_component(snapshot)];
    components.extend(gpu_components(snapshot));
    components.push(memory_component(snapshot));
    components.push(motherboard_component(snapshot));
    components.extend(disk_components(snapshot));
    for (kind, label, key) in [
        ("cooling", "CPU COOLING", "cpu_cooler"),
        ("case", "CHASSIS", "case"),
        ("power", "POWER SUPPLY", "power_supply"),
    ] {
        let description = snapshot.configured(key);
        if matches!(description, "" | "not set") {
            continue;
        }
        let brand = resolve_brand(kind, &[description]);
        components.push(Component {
            kind: kind.into(),
            label: label.into(),
            vendor: brand.name.clone(),
            model: strip_brand(description, brand),
            identifiers: vec![description.into()],
            compact: false,
            ..Component::default()
        });
    }
    components.extend(portable_power_components(snapshot));
    components
}
